tests/functional/test_package.py

```diff
@@ -23,6 +31,257 @@ def index():
+
+    @property
+    def calls(self):
+        return self._calls
+
+
+class PipSideEffect(object):
+    def __init__(self, filename, dirarg='--dest', consume=True):
+        self._filename = filename
+        self._package_name = filename.split('-')[0]
+        self._dirarg = dirarg
+        self._consume = consume
+
+    def _build_fake_whl(self, filepath):
+        if not os.path.isfile(filepath):
+            with zipfile.ZipFile(filepath, 'w') as z:
+                z.writestr('%s/placeholder' % self._package_name, b'')
+
+    def _build_fake_sdist(self, filepath):
+        # tar.gz is the same no reason to test it here as it is tested in
+        # unit.deploy.TestSdistMetadataFetcher
+        assert filepath.endswith('zip')
+        components = os.path.split(filepath)
+        prefix, filename = components[:-1], components[-1]
+        directory = os.path.join(*prefix)
+        filename_without_ext = filename[:-4]
+        pkg_name, pkg_version = filename_without_ext.split('-')
+        builder = FakeSdistBuilder()
+        builder.write_fake_sdist(directory, pkg_name, pkg_version)
+
+    def execute(self, args):
+        """Generate the file in the target_dir."""
+        if self._dirarg:
+            target_dir = None
+            for i, arg in enumerate(args):
+                if arg == self._dirarg:
+                    target_dir = args[i+1]
+            if target_dir:
+                filepath = os.path.join(target_dir, self._filename)
+                if filepath.endswith('.whl'):
+                    self._build_fake_whl(filepath)
+                else:
+                    self._build_fake_sdist(filepath)
+        return self._consume
+
+
+@pytest.fixture
+def osutils():
+    return OSUtils()
+
+
+@pytest.fixture
+def pip_runner():
+    pip = FakePip()
+    pip_runner = PipRunner(pip)
+    return pip, pip_runner
+
+
+class TestDependencyBuilder(object):
+    def _write_requirements_txt(self, contents, directory):
+        filepath = os.path.join(directory, 'requirements.txt')
+        with open(filepath, 'w') as f:
+            f.write(contents)
+
+    def test_can_get_whls_all_manylinux(self, tmpdir, osutils, pip_runner):
+        reqs = ['foo', 'bar']
+        pip, runner = pip_runner
+        pip.add_side_effect('download', [
+            PipSideEffect('foo-1.2-cp36-cp36m-manylinux1_x86_64.whl'),
+            PipSideEffect('bar-1.2-cp36-cp36m-manylinux1_x86_64.whl')
+        ])
+
+        appdir = str(_create_app_structure(tmpdir))
+        self._write_requirements_txt('\n'.join(reqs), appdir)
+        builder = DependencyBuilder(osutils, runner)
+        site_packages = builder.build_site_packages(appdir)
+        installed_packages = os.listdir(site_packages)
+        for req in reqs:
+            assert req in installed_packages
+
+    def test_can_get_whls_mixed_compat(self, tmpdir, osutils, pip_runner):
+        reqs = ['foo', 'bar', 'baz']
+        pip, runner = pip_runner
+        pip.add_side_effect('download', [
```

Comment:
So it looks like download is really the only command used for side effects. Would it be possible to simplify to something like this:
```python
pip.packages_to_download([
    'foo-1.0-cp36-none-any.whl',
    'bar-1.2-cp36-cp36m-manylinux1_x86_64.whl',
    'baz-1.5-cp36-cp36m-linux_x86_64.whl'
])
```
Ideally I would like to not have to worry about the `PipSideEffect` object